### PR TITLE
target/xilinx: Add block design flow

### PR DESCRIPTION
**Pull request by CyrilKoe**
The ethernet IP is quite complex and cannot be implemented by hand on FPGA, i.e. by instantiating Xilinx IP within the source files in System Verilog.

To solve this issue, we add two bitstream _flavors_:
- **Vanilla:** IPs integrated  by hand
- **BD (block design):** IPs integrated by block design

--- 

## Summary of changes

* The flow has gone through major changes to allow adding new flavor or boards easily.
* Safety island was bumped to change default burst mode (for compatibility with xilinx ips)
* A PCIe IP has been added but not tested yet.
* CI have been adapted to the new fpga flow

### CODE REVIEW

File: Bender.yml [hunk 1]
```diff
@@ -15,7 +15,7 @@ dependencies:
   cheshire:           { git: https://github.com/pulp-platform/cheshire.git,             rev: 0c95210cf242c384fafe3019e84b8974c3ff1e92 } # branch: aottaviano/carfield
   hyperbus:           { git: https://github.com/pulp-platform/hyperbus.git,             rev: f039e601c8b6590181734e6d26ff8b77aa380412 } # branch: chi/add_fsm_with_Tcsh
   dyn_mem:            { git: https://github.com/pulp-platform/dyn_spm.git,              rev: 480590062742230dc9bd4050358a15b4747bdf34 } # branch: main
-  safety_island:      { git: git@iis-git.ee.ethz.ch:carfield/safety-island.git,         rev: d0771928a2b850272c7a3d1b2a4a767bfb7c30a9 } # branch: michaero/carfield
+  safety_island:      { git: git@iis-git.ee.ethz.ch:carfield/safety-island.git,         rev: df98fe252cd8b7abbc1738002603680886a73959 } # branch: ck/incr_burst
```
Comment (CyrilKoe) [inline Bender.yml:18]: naming to be updated after merge request

File: README.md [hunk 2]
```diff
@@ -1,9 +1,9 @@
 # Carfield
```
Comment (CyrilKoe) [inline README.md:1]: The readme went though major changes due to automatic scripts of mkdocs, the removed lines are not lost but are now in the docs

File: bender-xilinx.mk [hunk 3]
```diff
@@ -21,6 +21,11 @@ $(eval $(call check_enable_island,GEN_PULP_CLUSTER))
 $(eval $(call check_enable_island,GEN_SAFETY_ISLAND))
```
Comment (CyrilKoe) [inline bender-xilinx.mk:21]: I recommend reading the documentation first to understand the new ext_jtag parameter

File: docs/img/arch.svg
Comment (CyrilKoe) [inline docs/img/arch.svg:1]: Empty file as placeholder for squeleton

File: hw/carfield.sv [hunk 4]
```diff
@@ -25,7 +25,7 @@ module carfield
   parameter islands_cfg_t IslandsCfg = carfield_pkg::IslandsCfgDefault,
   parameter int unsigned HypNumPhys  = 2,
   parameter int unsigned HypNumChips = 2,
-`ifdef NO_HYPERBUS // bender-xilinx.mk
```
Comment (CyrilKoe) [inline hw/carfield.sv:28]: Renamed with GEN for uniformity with other IPs (reminder we take the negative here for a hyperbus included by default)

File: sw/sw.mk [hunk 5]
```diff
@@ -145,9 +148,15 @@ mibench-automotive-susan: automotive-susan
 # GPT Linux image #
 ###################
 
+CAR_SW_DISK_SIZE := 16M
```
Comment (CyrilKoe) [inline sw/sw.mk:151]: One might make it larger to hold openmp libraries for instance

File: sw/sw.mk [hunk 6]
```diff
@@ -35,6 +35,9 @@ car-sw-libs: $(CAR_SW_LIBS)
 
 # Compilation
 
+carfield_%.dtb: carfield_%.dts
```
Comment (CyrilKoe) [inline sw/sw.mk:38]: Different rule than cheshire in order to add the include flag

File: target/xilinx/flavor_bd/constraints/carfield_ip.xdc [hunk 7]
```diff
@@ -0,0 +1,443 @@
+
```
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/constraints/carfield_ip.xdc:1]: Basically a copy paste of the vanilla carfield.xdc with the hierarchy modified for the block design

File: target/xilinx/flavor_bd/constraints/vcu128.xdc [hunk 8]
```diff
@@ -0,0 +1,1758 @@
+# VIOs are asynchronous
```
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/constraints/vcu128.xdc:1]: again copy paste to resolve path changes compared to vanilla

File: target/xilinx/flavor_bd/constraints/vcu128_ext_jtag.xdc
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/constraints/vcu128_ext_jtag.xdc:1]: add jtag constraints individually

File: target/xilinx/flavor_bd/constraints/vcu128.xdc
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/constraints/vcu128.xdc:1]: again copy paste to resolve path changes compared to vanilla

File: target/xilinx/flavor_bd/scripts/carfield_bd_vcu128.tcl
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/scripts/carfield_bd_vcu128.tcl:1]: auto generated file

File: target/xilinx/flavor_bd/scripts/run_carfield_ip.tcl
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/scripts/run_carfield_ip.tcl:1]: package the carfield ip

File: target/xilinx/flavor_bd/src/carfield_xilinx_ip.v [hunk 9]
```diff
@@ -0,0 +1,466 @@
...
+  inout  [`HypNumPhys-1:0][`HypNumChips-1:0] pad_hyper_csn,
+  inout  [`HypNumPhys-1:0]                   pad_hyper_ck,
+  inout  [`HypNumPhys-1:0]                   pad_hyper_ckn,
+  inout  [`HypNumPhys-1:0]                   pad_hyper_rwds,
+  // inout  [`HypNumPhys-1:0]                   pad_hyper_reset,
+  inout  [`HypNumPhys-1:0][7:0]              pad_hyper_dq,
+
+  // MASTER AXI DRAM
+
+(* X_INTERFACE_INFO = "xilinx.com:interface:aximm:1.0 dram_axi ARADDR" *) (* X_INTERFACE_PARAMETER = "FREQ_HZ 50000000" *)
```
Comment (CyrilKoe) [inline target/xilinx/flavor_bd/src/carfield_xilinx_ip.v:52]: We cannot use parameters on the interface of the IP

File: target/xilinx/flavor_vanilla/scripts/run.tcl [hunk 10]
```diff
@@ -2,21 +2,29 @@
 # Solderpad Hardware License, Version 0.51, see LICENSE for details.
 # SPDX-License-Identifier: SHL-0.51
 #
-# Author: Florian Zaruba <zarubaf@iis.ee.ethz.ch>
```
Comment (CyrilKoe) [inline target/xilinx/flavor_vanilla/scripts/run.tcl:5]: Considering the huge flow changes

File: target/xilinx/xilinx/common.mk
Comment (CyrilKoe) [inline target/xilinx/xilinx/common.mk:1]: renamed (xilinx_ips)

**Comment by CyrilKoe** [comment]
@alex96295 Docs were removed, I will remove the draft after CI finishes

File: target/xilinx/flavor_bd/src/carfield_xilinx_ip.v [hunk 11]
```diff
@@ -0,0 +1,466 @@
...
+    .pad_hyper_csn,
+    .pad_hyper_ck,
+    .pad_hyper_ckn,
+    .pad_hyper_rwds,
+    // .pad_hyper_reset,
+    .pad_hyper_dq,
+
+    // Dram axi
+
+    .dram_axi_m_axi_awid   (dram_axi_m_axi_awid   ),
```
Comment (alex96295) [inline target/xilinx/flavor_bd/src/carfield_xilinx_ip.v:332]: minor: you could skip connection here `.dram_axi_m_axi_awid,` since port and connection signal have the same name; less error prone

File: target/xilinx/xilinx_ips/common.mk [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+#
+# Artifacts management (IIS internal)
+#
+
+ifeq ($(XILINX_USE_ARTIFACTS),1)
+
+# Note: We do not use Memora as it is bound to Git versionning
+# and not standalone on files hash / environment variables
+ARTIFACTS_PATH=/usr/scratch2/wuerzburg/cykoenig/memora/carfield
```
Comment (alex96295) [inline target/xilinx/xilinx_ips/common.mk:40]: can we set this as env variable in the CI script so that we do not have to write down the machine's path explicitly?

**Comment by alex96295** [comment]
@CyrilKoe thanks for the great work

I went through it and only have minor comments. Before we merge, I have an observation:

I think the current directory structure inside the `target/` folder could be improved:
* The two flavors `vanilla` and `bd` are not currently reflected in the directories names; I suggested as an alternative to rename `xilinx/` in `xilinx_vanilla/` and then add both `xilinx_vanilla/` and `xilinx_bd/` under a `xilinx/` directory within `target/` (see suggested tree below)
* The readme inside the current `xilinx/` directory could be removed, or moved to the top-level `xilinx/` folder if you apply the suggestions in the step above, pointing to the documentation
* For both flavors, we rely on ooc implementation of IPs: in the vanilla design, the various xilinx wizards; in the block design, carfield itself. So I would keep symmetry with an additional `ips/` folder where the ooc designs are kept, for both favors
  
Here follows a tree with what I had in mind for the directory structure (plus comments)

```tree
target
├── synth
│   └── carfield_synth_wrap.sv
└── xilinx // wrap xilinx_bd/ and xilinx_vanilla in xilinx top directory
    ├── xilinx_bd
    │   ├── constraints // keep ooc carfield constraints at top since they are typically reused
    │   │   ├── carfield_ip.xdc
    │   │   ├── ooc_carfield_ip.xdc
    │   │   ├── vcu128.xdc
    │   │   └── zcu102.xdc
    │   ├── ips // add carfield in a ips/ folder as ooc IP
    │   │   └── xlnx_carfield
    │   │       ├── scripts
    │   │       │   └── run.tcl // script is renamed run.tcl since the tree structure makes you understand where you are already (i.e., implementing carfield ooc)
    │   │       └── src
    │   │           ├── carfield_xilinx_ip.v
    │   │           └── carfield_xilinx.sv
    │   ├── scripts
    │   │   ├── carfield_bd_vcu128.tcl
    │   │   └── run.tcl
    │   └── xilinx_bd.mk
    └── xilinx_vanilla
        ├── constraints
        │   ├── carfield.xdc
        │   ├── vcu128.xdc
        │   └── zcu102.xdc
        ├── ips //renamed 'xilinx/' to 'ips/' as discussed above 
        │   ├── common.mk
        │   ├── xlnx_clk_wiz
        │   │   ├── Makefile
        │   │   └── tcl
        │   │       └── run.tcl
        │   ├── xlnx_mig_ddr4
        │   │   ├── Makefile
        │   │   └── tcl
        │   │       └── run.tcl
        │   └── xlnx_vio
        │       ├── Makefile
        │       └── tcl
        │           └── run.tcl
        ├── scripts
        │   ├── flash_spi.tcl
        │   ├── overrides.sh
        │   ├── program.tcl
        │   ├── prologue.tcl
        │   ├── run.tcl
        │   └── write_cfgmem.tcl
        ├── src
        │   ├── carfield_top_xilinx.sv
        │   ├── cdc_dst_axi_err.sv
        │   ├── dram_wrapper.sv
        │   ├── fan_ctrl.sv
        │   ├── overrides
        │   │   └── tc_clk_xilinx.sv
        │   └── phy_definitions.svh
        └── xilinx_vanilla.mk

``` 

Could you please have a look at it and provide your opinion? :)

**Comment by CyrilKoe** [comment]
@alex96295 thanks for the review and the feedback, about the directory structure here are the reasons motivating it:
- __Makefiles:__ In order to have a normalized flow and reduce phonies (to only `car-xil-all`), I needed to have a common `xilinx.mk` that includes both flavoured makefiles before describing the common rules **at the end** (to allow using flavor-dependant variables as target dependencies for these rules). It could be that `xilinx_vanilla.mk` includes the other flavors but it would make it long and hard to read :  
"Common variables, Vanilla variables, Vanilla targets, Incude other flavors, Flavor dependant Variables, Common targets"  
- __Future flavors:__ If a contributor decides to add a new flavor (let's imagine a template-generated RTL top level - as Occamy - that would be very different from the vanilla flavor while still not relying on a block design). It would also require `xilinx_ips` and this is why I consider them as a "common" more than a "vanilla" feature. The IP config file would then contain `if FLAVOR==vanilla_2` if reconfiguration is needed. But appart from `run.tcl` everything is common and should not be associated with vanilla only to avoid duplication.
- __Extension to other projects:__ one can copy this emulation flow in a new project without the vanilla flavor if they only wish to implement a block design (which is easier in most of the cases, and also the Vivado legit way to do). Vanilla directory should not be required.

This is why I think the top level Xilinx directory, is necessary and should contain common scripts (`flash`; `program`; `overrides`; `write_cgmem`); common srcs (like the `axi_cdc_err_slave` _it should maybe even be added helow `/hw` at the root of the repository_) and the common `xilinx.mk`. But:

- `flavor_vanilla` and `flavor_bd` could be renamed in `xilinx_flavor_vanilla` and `xilinx_flavor_bd`. For more clarity if one do not care about the fpga flow.
- A `user_ips` folder can be created below the `flavor_bd` to split constraints and scripts between block design related and carfield ip related. I would recommend keeping it as `user_ips` as the goal is quite different from `xilinx_ip`. (_Note:_ why having carfield_ip.xdc one directory above in your tree?)  
_Improvement:_ `users_ip` could even be actually at the root as it would be a common between different flavor of block designs, but this may be exagerated in the current state...
- _Improvement:_ `carfield.xdc` could be a common by taking a variable to override each path with a given prefix.
- The readme could be removed as there is the doc. As any flavor will share anyways the same `carfield_soc.sv`)

Precision, in my view new flavors should be added like this:

```
├── xilinx
│    ├── scripts
│    ├── constraints
│    ├── src
│    ├── xilinx_flavor_vanilla
│    ├── xilinx_flavor_vanilla_generated
│    ├── xilinx_flavor_bd_ethernet
│    ├── xilinx_flavor_bd_pcie_device
│    ├── xilinx.mk
```

Of course these may not be relevant for Carfield but may be for future SoCs.

File: target/xilinx/xilinx_ips/common.mk [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+
+all: load-artifacts $(PROJECT).xpr save-artifacts
+
+# Build IP
+xlnx_%.xpr:
+	$(vivado_env) $(VIVADO) -mode batch -source tcl/run.tcl
+
+save-artifacts:
+
+load-artifacts:
```
[the remark this answers is not in the document]
Comment (CyrilKoe) [reply target/xilinx/xilinx_ips/common.mk:15]: This was to avoid override the target `all` as done [before](https://github.com/pulp-platform/carfield/blob/main/target/xilinx/xilinx/common.mk#L18).
If it's bad practice I can remove it

File: target/xilinx/xilinx_ips/common.mk (hunk 12 above)
Comment (CyrilKoe) [reply target/xilinx/xilinx_ips/common.mk:40]: Sure

**Comment by alex96295** [comment]
> @alex96295 thanks for the review and the feedback, about the directory structure here are the reasons motivating it:
> 
>     * **Makefiles:** In order to have a normalized flow and reduce phonies (to only `car-xil-all`), I needed to have a common `xilinx.mk` that includes both flavoured makefiles before describing the common rules **at the end** (to allow using flavor-dependant variables as target dependencies for these rules). It could be that `xilinx_vanilla.mk` includes the other flavors but it would make it long and hard to read :
>       "Common variables, Vanilla variables, Vanilla targets, Incude other flavors, Flavor dependant Variables, Common targets"
> 
>     * **Future flavors:** If a contributor decides to add a new flavor (let's imagine a template-generated RTL top level - as Occamy - that would be very different from the vanilla flavor while still not relying on a block design). It would also require `xilinx_ips` and this is why I consider them as a "common" more than a "vanilla" feature. The IP config file would then contain `if FLAVOR==vanilla_2` if reconfiguration is needed. But appart from `run.tcl` everything is common and should not be associated with vanilla only to avoid duplication.
> 
>     * **Extension to other projects:** one can copy this emulation flow in a new project without the vanilla flavor if they only wish to implement a block design (which is easier in most of the cases, and also the Vivado legit way to do). Vanilla directory should not be required.
> 
> 
> This is why I think the top level Xilinx directory, is necessary and should contain common scripts (`flash`; `program`; `overrides`; `write_cgmem`); common srcs (like the `axi_cdc_err_slave` _it should maybe even be added helow `/hw` at the root of the repository_) and the common `xilinx.mk`. But:
> 
>     * `flavor_vanilla` and `flavor_bd` could be renamed in `xilinx_flavor_vanilla` and `xilinx_flavor_bd`. For more clarity if one do not care about the fpga flow.
> 
>     * A `user_ips` folder can be created below the `flavor_bd` to split constraints and scripts between block design related and carfield ip related. I would recommend keeping it as `user_ips` as the goal is quite different from `xilinx_ip`. (_Note:_ why having carfield_ip.xdc one directory above in your tree?)
>       _Improvement:_ `users_ip` could even be actually at the root as it would be a common between different flavor of block designs, but this may be exagerated in the current state...
> 
>     * _Improvement:_ `carfield.xdc` could be a common by taking a variable to override each path with a given prefix.
> 
>     * The readme could be removed as there is the doc. As any flavor will share anyways the same `carfield_soc.sv`)
> 
> 
> Precision, in my view new flavors should be added like this:
> 
> ```
> ├── xilinx
> │    ├── scripts
> │    ├── constraints
> │    ├── src
> │    ├── xilinx_flavor_vanilla
> │    ├── xilinx_flavor_vanilla_generated
> │    ├── xilinx_flavor_bd_ethernet
> │    ├── xilinx_flavor_bd_pcie_device
> │    ├── xilinx.mk
> ```
> 
> Of course these may not be relevant for Carfield but may be for future SoCs. 

Yes, so first of all my apologies because I based my `tree` on `ck/xilinx_bd` instead of `ck/xilinx_bd_wip` where you addressed some of the points I mentioned already. I agree we should try to maximize common code, I see in your branch you already did it.

I agree with calling ooc IPs from user `user_ips`, and if it is possible for you to implement it in this PR, that would be great. I left `carfield_ip.xdc` at the top in my tree because I thought that the constraints need to be re-included in the top level BD anyway from what I could recall, maybe it's not the case for this flow and we can keep it at `user_ip` level

I don't think that renaming the `flavor_*` directory is required as you are already inside the `xilinx` folder, so no need to prefix it again

For the carfield.xdc and override: up to you if you have time to do it now. I was noticing some overlapping between constraints and IMHO we should try to minimize those as much as possible

---
TL;DR: action items

1. differentiate between `user_ips` and `xilinx_ips` folders
2. Try to reduce overlap in constraints as much as possible

If you could address those above, then we are ready to merge for me :)

**Comment by CyrilKoe** [comment]
After reconsidering; `carfield_ip` could be included in `xilinx_ips`, but it would be easier then to change the `xilinx_ips` into a `.mk` solution rather than an independant `Makefile` system (what @paulsc96 recommended on Cheshire).
Like this I could access the common bender targets and other useful variables to buile the carfield ip.

I don't have time right now; but i'll do soonish it if you think it works.

**Comment by alex96295** [comment]
> After reconsidering; `carfield_ip` could be included in `xilinx_ips`, but it would be easier then to change the `xilinx_ips` into a `.mk` solution rather than an independant `Makefile` system (what @paulsc96 recommended on Cheshire). Like this I could access the common bender targets and other useful variables to buile the carfield ip.
> 
> I don't have time right now; but i'll do soonish it if you think it works.

of course, that works: using a makefrag is recommended. For what concerns adding it to `xilinx_ips`, we said that's a kind of common place for the various flows, while carfield ooc is just for bd design.  So maybe a standalone folder with its own makefrag would help. I liked the `user_ips` idea because they are two logically separated things

This won't prevent changing the xilinx_ips in a makefrag solution as well

**Comment by alex96295** [comment]
I see, thanks :)
Just one thing, what do you mean by:

> to not reproduce the complexity

**Comment by CyrilKoe** [comment]
Just that xilinx_ips.mk is pretty complex, so having another complex user_ips.mk would be a lot. But it makes sense that carfield_ip is compiled with the same makefile as the outputs are exactly the same (ip.xci)

**Comment by alex96295** [comment]
> Just that xilinx_ips.mk is pretty complex, so having another complex user_ips.mk would be a lot. And it makes sense that carfield_ip is compiled with the same makefile as the outputs are exactly the same (ip.xci)

Yes, I agree

**Comment by alex96295** [comment]
If fpga mapping passes on CI, I guess that a rebase on main will be the last step, and then we can merge :)

**Comment by CyrilKoe** [comment]
Rebased, [pipeline](https://iis-git.ee.ethz.ch/github-mirror/carfield/-/pipelines/80763) running.
Thanks :D
